app/controllers/v0/sessions_controller.rb

```diff
@@ -103,18 +103,30 @@ def saml_callback
     def persist_session_and_user
       saml_attributes = SAML::User.new(@saml_response)
       existing_user = User.find(saml_attributes.user_attributes.uuid)
-      @current_user = User.new(saml_attributes.to_hash)
-      @current_user.last_signed_in = if saml_attributes.changing_multifactor? && existing_user.present?
-                                       existing_user.last_signed_in
-                                     else
-                                       Time.current.utc
-                                     end
+      user_identity = UserIdentity.new(saml_attributes.to_hash)
+      @current_user = init_new_user(user_identity, existing_user, saml_attributes.changing_multifactor?)
 
-      StatsD.increment(STATSD_LOGIN_NEW_USER_KEY) unless existing_user.present?
-      existing_user.destroy if existing_user.present?
+      if existing_user.present?
+        existing_user&.identity&.destroy
```

Comment:
Could this be rolled into `user.destroy`?

Reply:
haha eventually yes, i think that would be nice... maybe i could figure out how to introduce an accepts_nested_attributes_for type of mechanism on RedisStore too. But putting destroy on User would have to happen with after_destroy callbacks, which currently don't exist, but could easily be added.

Reply:
actually... i could just override the destroy method, have identity destroyed and then call super, or would you prefer it be done in a callback?

Reply:
leaving as is for now -- this could all be simplified elsewhere, not crazy about introducing callbacks right now because I strongly feel that callbacks should be used only when the logic refers to state internal to the object. In this case, it's a different object. I'm  not a big fan of "dependent destroy" and things like it. If anything all this user persistance stuff belongs in its own class as evidenced by the cyclomatic complexity warnings.

Reply:
I was thinking override and super